Initial release

fillup opened the pull request.

fillup committed:
initial version

fillup committed:
merge master into develop

Baggerone commented on src/EmailServiceClient.php:
```diff
@@ -0,0 +1,258 @@
+    }
+    
+    /**
+     * Create a user with the given information.
```
This needs to be corrected, right?

fillup replied:
yep, thanks. fixed.

scrutinizer-auto-fixer committed:
Scrutinizer Auto-Fixes

This commit consists of patches automatically generated for this project on https://scrutinizer-ci.com

fillup committed:
Merge pull request #2 from silinternational/scrutinizer-patch-1

Scrutinizer Auto-Fixes

forevermatt commented on features/response/ResponseContext.php:
```diff
@@ -0,0 +1,159 @@
+    public function theResultShouldNotContainUserInformation()
+    {
+        if (is_array($this->result)) {
+            foreach ($this->userInfoFields as $fieldName) {
```
Oops... looks like `$this->userInfoFields` was accidentally removed.
We may want a line something like this at the beginning of this function:
```
Assert::assertNotEmpty($this->userInfoFields);
```

forevermatt commented on features/request/request.feature:
```diff
@@ -0,0 +1,55 @@
+    When I create the emailServiceClient
+    Then an InvalidArgument exception will be thrown
+
+  Scenario: Checking client with the assert_valid_ip not given and the trusted_ip_ranges value missing
```
I don't really understand what this scenario is trying to prove about the code (based on the Given/When/Then steps it uses). Is it proving that, if not told whether to assert that the IP address is trusted (and thus doing so by default) and not given a range of trusted IP addresses, an InvalidArgumentException is thrown?

If so, it might be more readable to write it as...
```
  Scenario: Requiring a trusted_ip_ranges value when assert_valid_ip is missing
    Given I have provided a baseUri
      And I have not indicated whether the email service ip should be validated
      And I have not provided a list of trusted IP ranges
    When I create the emailServiceClient
    Then an InvalidArgument exception will be thrown
```

forevermatt replied:
Since having hard-to-read tests can lead to its own set of problems, I can update this Behat test as written here after this PR is merged. No reason to hold things up for that refactor, though.

Baggerone replied:
Note: 
The assertValidIp is true by default.  So, if no assert_valid_ip config is available, then the ip will be evaluated against the trusted ones.

forevermatt commented on run-tests.sh:
```diff
@@ -0,0 +1,13 @@
+rc=$?; if [[ $rc != 0 ]]; then exit $rc; fi
+
+# Run the feature tests
+vendor/bin/behat --stop-on-failure
```
Could this be updated to use `runny`? Same with the `composer install` line, above.

fillup replied:
we need runny for commands that bash will ignore errors on and proceed to next command, but the last command in the file does not need runny so we can see its full output and its exit code will stop a build if it fails anyway

forevermatt replied:
Oh, does the return value / output of the last command in a bash script end up being the return value / output when the bash script itself is run? If so, is there anything else runny would gain us that is worth having here (like logging to syslog)?

fillup committed:
remove unused function and update comments

forevermatt commented on src/EmailServiceClient.php:
```diff
@@ -0,0 +1,260 @@
+
+    /*
+     * Validates the config values for ASSERT_VALID_IP_CONFIG and
+     *   ASSERT_VALID_IP_CONFIG
```
Should this `ASSERT_VALID_IP_CONFIG` have `TRUSTED_IPS_CONFIG`?

fillup replied:
yeah

forevermatt commented on src/EmailServiceClient.php:
```diff
@@ -0,0 +1,260 @@
+     * @return array|null An array of user information (if valid), or null.
+     * @throws Exception
+     */
+    public function authenticate(string $username, string $password)
```
I think this whole function can be deleted, as it seems to be a carryover from the ID Broker PHP Client that isn't applicable here.

forevermatt commented on src/EmailServiceClient.php:
```diff
@@ -0,0 +1,260 @@
+    }
+    
+    /**
+     * Create a user with the given information.
```
Looks like the doc. comment for this method needs to be updated.

forevermatt commented on src/EmailServiceClient.php:
```diff
@@ -0,0 +1,260 @@
+        $result = $this->emailInternal($config);
+        $statusCode = (int)$result['statusCode'];
+        
+        if ($statusCode === 200) {
```
Do we want to allow any `2xx` response? We had to do that for siteStatus (see [line 204](https://github.com/silinternational/email-service-php-client/pull/1/files#diff-32b9ee4d143c66b3fa426dd636c436d8R204), below).

fillup replied:
sure

fillup committed:
removed unused behat context methods

fillup committed:
pr feedback updates

fillup committed:
mend
